### ISSUE TITLE
Round-robin load balance fairness with slow queue subscribers

### ISSUE THREAD

**Issue by justinfx** [nats-io/nats-server#878]
I've been tracking down an issue in my application where my workers in a queue subscription are not being utilised properly. The structure of my application is:

* Queued subscription of workers that receive task requests, do some processing, reply with the result and then try to get the next available task again
* Clients send a request to process a task and wait for the reply

When I have 3 idle workers I can send 3 requests in parallel which instruct the workers to sleep for 5 seconds and then reply. I would expect to see behaviour where all 3 idle workers are being utilised for the 3 requests. But instead it seems like requests can pile up on a busy worker, making the work inefficient.  

I've been given information about the nature of subscription queues having no concept of whether the receivers are actually able to handle the next message:
https://natsio.slack.com/archives/C069GSYFP/p1547221881169300

But this information is concerning because I now have more than one project which relies on subscription queues with workers that are doing time consuming work in their message handlers before accepting the next message. I now wonder if nats queues are efficient enough for these situations or if I have to create my own broker in the middle manages the queue and knows which worker is waiting for a message.

#### Versions of `gnatsd` and affected client libraries used:

gnatsd / embedded server 1.3.0
Go client library 1.6.0

#### OS/Container environment:

Ubuntu 16.04

#### Steps or code to reproduce the issue:

Please reference the following complete reproduction of the issue:
https://gist.github.com/justinfx/b0adb36e694ec03365da19b6bbf33c20
 
#### Expected result:

Each idle worker subscriber would receive one of the 3 available messages
 
#### Actual result:

2 workers receive a message and take time to process, while the 3rd worker remains idle. Once one of the other workers finishes its message handling it receives the 3rd message

Output from gnatsd:
[gnatsd_out.txt](https://github.com/nats-io/gnatsd/files/2753498/gnatsd_out.txt)

**Comment by justinfx** [comment]
Thanks for the reply on this, Derek. I'm told that using NATS streaming could solve my problem because it uses ACKs. In general should I make sure to use NATS streaming for queues like this where the handler on the subscription is slow and I am looking for more consistent load balancing to idle subscribers?

**Comment by derekcollison** [comment]
NATS Streaming does understand outstanding messages that have not been ack'd and uses that to select subscribers. This may be a better fit. For me I would try to understand what the responder's are doing that makes them slow (for instance if its IO, you can run many of them inside of a single process/subscriber).

**Comment by justinfx** [comment]
The issue is that they are cpu/memory constrained. It is being used as an arbitrary task subprocess queue. When a message has been received by a subscriber, it is intended for that worker process to only be executing that single task. Lets say the task is to call a subprocess that will convert 100 images and this takes 10 seconds. This worker is now busy for 10 seconds, yet there are other workers that are idle. So I would not want the next message to be enqueued to this busy workers pending buffer. There is nothing I can do about the time it takes for commands to execute (other than that I enforce max timeouts). But like you said, maybe this workload is not suited to the random scheduling used by NATS core queue subscribers. It ends up wasting idle resources.
I will try integrating a streaming service to the mix to see if this fixes my issue.

**Comment by derekcollison** [comment]
Try NATS Streaming with queue subscribers and max inflight of 1.

**Comment by justinfx** [comment]
**Update:** 

I rewrote the connection and message logic in my app to use an embedded streaming nats server. Because I found that the streaming server only supports pub/sub and no wildcard, I had to keep all the auxillary non-task-processing messages going through the underlying nats client connection, Then I set up the main task request process to do the following:

**Sender**
1. Create an inbox and subscribe through plain nats as a reply channel with auto unsubscribe 1
1. Encode the reply address into the task payload
1. Publish my task request through the streaming client
1. Receive reply

**Worker**
1. Subscribe through streaming client to queue, with `MaxInflight=1`, `SetManualAckMode=true` and `AckWait=30s` (to account for long processing). This would have to be much longer to account for very long tasks.
1. Receive a task, handle
1. Reply through plain nats client using task reply id
1. Ack

Strangely I found it to only improve the likelihood of scheduling N to N. Out of about 10 runs of sending 8 tasks to 8 idle subscribers I would see it schedule 7+1 about 1 or 2 times. I then reduced the problem using my previously posted repro test, but now with streaming and a standalone `stan` instance. For some reason I can not reproduce the scheduling problem when it is the reduced standalone example, yet I can consistently reproduce it in my embedded application, even when I confirm the settings all the same.

I'm going to keep investigating this, but I am finding it not exactly that easy to use nats.io as a reliably scheduled push/pull work queue for work that takes time. I forgot to mention another constraint for this application: One use case is that a client expects to break apart a simulation type task into N parts and have them all start together on N workers. So if they know they send 8 requests and only 7 start and the 8th one is sitting behind one of the busy workers in its queue, the whole client request stalls and times out.

I really love using nats to solve my distributed messaging needs so I want to give it my all to find a working solution to this. Do you think if I cant get streaming to solve this, that it would be a potentially workable solution if I had an application broker in the middle that receives all requests into its own queue, and then the workers use a request/reply pattern to ask the broker subject for work?

A second alternate idea I had was to have the workers auto unsubscribe after every message, and then subscribe when it finishes its task. I would only use this mode of operation when we expect the task types to be long running. Would this be really bad for the nats server?

**Comment by derekcollison** [comment]
We are looking closely at more traditional work queue scenarios for JetStream, the successor to NATS Streaming.

**Comment by justinfx** [comment]
Thanks. I looked up JetStream and found that it is now called Liftbridge? That will be great to have official support for different strategies of work queues.

Final update to this ticket... 

I managed to get a working solution after testing all the available options I had mentioned previously. No matter what I tried, I could not get the streaming server approach to consistently schedule to idle subscribers. It still hit a 20% case of scheduling the 8th concurrent message to one of the previous 7 busy subscribers. I also tried unsubscribing after receiving a message, but this was a terrible idea that I had since it just made acks fail with bad subscriptions. And it also failed in the core nats approach because it was still a race against messages being delivered.

The solution that works consistently was to introduce a broker subject between the requestor and the worker and put the request into an unbuffered chan in individual goroutine handlers. Then workers will make a request each time they want a new task, which will be a sync chan operation. This results in workers getting exactly one message until they are done and ready to ask for another. 

```
+------------------+     +------------+     +----------------+
| *Client*         |     | *Broker*   |     | *Worker*       |
| Publish()        |---> |  <-chan<-  |---> | Request()      |
|------------------+     +------------+     |----------------+
|                  |                        |                |
| Subscribe(reply) | <----------------------| Publish(reply) |
+------------------+                        +----------------+
```

**Comment by derekcollison** [comment]
LiftBridge is from another developer @tylertreat. JetStream will be NATS Streaming V2, so they should not be thought of as the same.

**Comment by derekcollison** [comment]
Did you set max in flight to 1 for the NATS streaming setup where no additional messages would be delivered until they were acked?

**Comment by ripienaar** [comment]
Over time these things even out, I am convinced you'll never see it balance evenly on the small scale you're looking at.

```
stream_replicator_copied_msgs{name="replicator_0",worker="replicator"} 1.117204e+06
stream_replicator_copied_msgs{name="replicator_1",worker="replicator"} 1.114914e+06
stream_replicator_copied_msgs{name="replicator_2",worker="replicator"} 1.115926e+06
stream_replicator_copied_msgs{name="replicator_3",worker="replicator"} 1.115933e+06
stream_replicator_copied_msgs{name="replicator_4",worker="replicator"} 1.114946e+06
stream_replicator_copied_msgs{name="replicator_5",worker="replicator"} 1.11594e+06
stream_replicator_copied_msgs{name="replicator_6",worker="replicator"} 1.114339e+06
stream_replicator_copied_msgs{name="replicator_7",worker="replicator"} 1.11636e+06
stream_replicator_copied_msgs{name="replicator_8",worker="replicator"} 1.115467e+06
stream_replicator_copied_msgs{name="replicator_9",worker="replicator"} 1.114198e+06
```

```
stream_replicator_copied_bytes{name="replicator_0",worker="replicator"} 8.07301724e+09
stream_replicator_copied_bytes{name="replicator_1",worker="replicator"} 8.057998094e+09
stream_replicator_copied_bytes{name="replicator_2",worker="replicator"} 8.064699705e+09
stream_replicator_copied_bytes{name="replicator_3",worker="replicator"} 8.065053196e+09
stream_replicator_copied_bytes{name="replicator_4",worker="replicator"} 8.057635788e+09
stream_replicator_copied_bytes{name="replicator_5",worker="replicator"} 8.06186233e+09
stream_replicator_copied_bytes{name="replicator_6",worker="replicator"} 8.051735716e+09
stream_replicator_copied_bytes{name="replicator_7",worker="replicator"} 8.065190224e+09
stream_replicator_copied_bytes{name="replicator_8",worker="replicator"} 8.062884557e+09
stream_replicator_copied_bytes{name="replicator_9",worker="replicator"} 8.051237738e+09
```

This with `MaxInflight(10)`

**Comment by justinfx** [comment]
@derekcollison , sorry for the confusion. I tried to look up into on JetStream and ended up seeing a blog post where they said it ended up becoming LiftBridge. Also, I did set max inflight specifically as the subscription option for the worker, and nothing specific on the streaming server. I could look at that option again I support.

@ripienaar evening out over time is not acceptible in my work queue requirements. Consider that there is a workflow that I want to run which can be broken up into 4 tasks that all connect and work on that workflow. A specific example is like an MPI process, where you need your 4 ranks to start up and connect and then all start working on the workflow. If I have 4 idle subscribers and send 4 messages at the same time, and 3 start up and 1 sits in the pending queue of a busy subscriber, then my workflow stalls until timeout. I specifically need to utilise my available workers, which neither core queue subscriptions nor streaming server (so far) can guarantee. 
When I switch to a model of my workers requesting the next task and coordinate my own request queue with a broker subject in the middle, I can guarantee that coordination. But this seems like a pretty traditional queuing model, something like one could get with zmq push/pull pattern, only with zmq I wouldn't have to maintain my own queue in the middle.

**Comment by derekcollison** [comment]
Closing this for now as stale, but if it needs to be re-opened please feel free to ping me.

**Comment by timje** [comment]
I have the same requirements as @justinfx (no idle subscribers when work is still in queue)
Is the broker subject that @justinfx described the only solution to ensure no idle subscribers?

**Comment by derekcollison** [comment]
NATS core is at most once, so queue subscribers are randomly selected and the message is sent with that QoS, meaning if they can't process for whatever reason that request is lost and the requestor should timeout and retry. If you want more intelligent queue subscriber selection that is semantically aware regarding number of outstanding messages use NATS streaming, which offers at least once semantics.

**Comment by justinfx** [comment]
I didn't have a consistent solution for this specific problem using nats streaming, but I am having great success with streaming on a more recent project. So it was probably just not suited to this queuing problem.

My solution of having a broker has been working great. Every idle consumer gets a message and I fully utilise my workers. It has some naive handling of workers going away after they have asked for another task and one is finally ready to give it. I'm timing out the sending of the task to the worker after a second and retrying it on another worker. But I could update it to ping the workers after they ask for a task and preemptively bail on them when they go away.

**Comment by puneetchelani** [comment]
@justinfx  We are facing the same issue with nats queues. Not sure if we can implement the unbuffered channels as our subscribers are written in python. 

Only solution I have is to manage the free subscribers in golang code and only publish to a free python subscriber on unique subject without using queue group. Issues here will be, we are sending multiple kinds of request to python and managing free subscribers in golang can be error prone.

@derekcollison do we have a solution where we have some load balancing mechanism other than random.

**Comment by justinfx** [comment]
@puneetchelani, I obviously don't know the details of your implementation, but regarding the way I have implemented it in my own application I can't imagine it would matter which language the worker subscribers are implemented in. When a subscriber is ready to accept new work it will send a request, and then go into a wait mode for a new task to be sent on its unique reply subject.
On the broker side, that "worker ready" request spins off a goroutine which then waits for the internal task channel to produce a new task. It will then forward it to that unique reply subject for the worker. So if 10 workers report in as being free, I have 10 goroutines pulling from the task request channel to service each unique worker. Should the worker go away after a task is delivered, it just requeues the task and bails. The part I intend to improve is actively polling the worker while waiting for the next task so I can bail on it earlier and not have to requeue.
Basically I have just turned it into a manual push/pull with the broker handling all incoming task requests. I imagine arbitrary workers (any language) could participate.

**Comment by puneetchelani** [comment]
@justinfx I think it all make sense. If we dont have a inbuilt solution, managing queueing and instant delivery  ourselves is a way out.
Actually, I was also thinking to implement something similar, where all requests will be pushed to workers by publisher who are waiting on different subjects without a queue group and I have to maintain list of all free subscribers with different subjects in code which I can do easily in my app.

**Comment by derekcollison** [comment]
So just to be clear from my side, this is just a plain worker queue problem that is not directly addressed by NATS core or NATS streaming? Meaning you just want worker queue semantics.

**Comment by justinfx** [comment]
@derekcollison yes that's correct. Nats is amazing, but it just lacks the ability to provide a high level push/pull worker queue that is based on worker availability. And so we must implement our own using the req/reply primitives. I imagine nats core could easily integrate these concepts into another type of queue group that was based on
* workers publishing they are ready with unique reply subject
* timeout option for how long a worker can take to accept the msg being sent in its unique subject 
* nats server knowing workers are still waiting on their unique reply subjects to mitigate need for requeue (republish into queue group)

**Comment by derekcollison** [comment]
Originally I thought worker queues were solved problem and should not be duplicated in NATS. With 2.0 and the future JetStream we will provide streaming and worker queues. But will be a bit post 2.0 release on June 10th.

**Comment by justinfx** [comment]
That's great to hear about it being part of the 2.0 roadmap. While it is a solved problem, it just means one needs a second message bus or external solution (ala redis).
Thanks!

**Comment by derekcollison** [comment]
Understood. JetStream will be pretty slick and handle multiple cases like streaming and MQs with one tech.
